sql/core/src/main/scala/org/apache/spark/sql/artifact/ArtifactManager.scala

```diff
@@ -279,6 +284,45 @@ class ArtifactManager(session: SparkSession) extends Logging {
     loader
   }
 
+  private[sql] def clone(newSession: SparkSession): ArtifactManager = {
+    val newArtifactManager = new ArtifactManager(newSession)
+    if (artifactPath.toFile.exists()) {
+      FileUtils.copyDirectory(artifactPath.toFile, newArtifactManager.artifactPath.toFile)
+    }
+    val oldBlockManager = session.sparkContext.env.blockManager
+    val newBlockManager = newSession.sparkContext.env.blockManager
+    val newBlockIds = cachedBlockIdList.asScala.map { blockId =>
+      val newBlockId = CacheId(sessionUUID = newSession.sessionUUID, hash = blockId.hash)
+      transferBlock(blockId, newBlockId, oldBlockManager, newBlockManager)
+    }
+    newArtifactManager.cachedBlockIdList.addAll(newBlockIds.asJava)
+    newArtifactManager.jarsList.addAll(jarsList)
+    newArtifactManager.pythonIncludeList.addAll(pythonIncludeList)
+    newArtifactManager
+  }
+
+  private def transferBlock(
```

Comment:
Rename to `copyBlock` rather than "transfer"?

And nit: we can put this as a private method in the companion object since its stateless